[inductor] Allow view to take dtype as its input

desertfire opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #88588

Summary: as title

Test Plan:
```
PYTORCH_TEST_WITH_INDUCTOR=1 python test/test_ops.py -k test_non_standard_bool_values
```

cc @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @chunyuan-w @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10

desertfire committed:
[inductor] Allow view to take dtype as its input

Summary: as title

Test Plan:
```
PYTORCH_TEST_WITH_INDUCTOR=1 python test/test_ops.py -k test_non_standard_bool_values_argwhere_cuda_bool
```

[ghstack-poisoned]

eellison reviewed (commented):
View doesn't do dtype conversion, it's basically `reinterpret_cast`. See:
```
>>> torch.rand([2]).view(torch.float16).shape
torch.Size([4])
```

ngimel commented on torch/_inductor/lowering.py:
```diff
@@ -690,10 +690,14 @@ def inner_fn(index):
+def view(x, sizes_or_dtype):
+    # view could also take dtype as its parameter
+    if isinstance(sizes_or_dtype, torch.dtype):
+        return to_dtype(x, sizes_or_dtype)
```
This is not correct semantics (and would be revealed by a test), correct dtype view semantics is bitcast:
```
In [11]: a=torch.randn(2,3)

In [12]: a
Out[12]: 
tensor([[ 1.1923, -1.2699, -0.8661],
        [ 0.7665,  0.1736,  0.9688]])

In [13]: a.view(dtype=torch.int)
Out[13]: 
tensor([[ 1066965979, -1079866187, -1084376557],
        [ 1061435139,  1043447875,  1064829971]], dtype=torch.int32)

In [14]: a.to(dtype=torch.int)
Out[14]: 
tensor([[ 1, -1,  0],
        [ 0,  0,  0]], dtype=torch.int32)
```

desertfire replied:
Is there any util function in `prim` for this kind of conversion?

ngimel replied:
No, I don't think so, and in any case we'd need to add triton codegen support for it.

desertfire committed:
Update on "[inductor] Allow view to take dtype as its input"


Summary: as title

Test Plan:
```
PYTORCH_TEST_WITH_INDUCTOR=1 python test/test_ops.py -k test_non_standard_bool_values
```

[ghstack-poisoned]

desertfire commented:
I took a shortcut to support a conversion with the same dtype length only. The more general conversion doesn't seem to have a high priority for now.

ngimel reviewed (commented):
Awesome, cool that triton supports bitcast. Can you add an inductor test so that we don't rely on general pytorch suite?

desertfire committed:
Update on "[inductor] Allow view to take dtype as its input"


Summary: as title

Test Plan:
```
PYTORCH_TEST_WITH_INDUCTOR=1 python test/test_ops.py -k test_non_standard_bool_values
```

[ghstack-poisoned]

desertfire committed:
Update on "[inductor] Allow view to take dtype as its input"


Summary: as title

Test Plan:
```
PYTORCH_TEST_WITH_INDUCTOR=1 python test/test_ops.py -k test_non_standard_bool_values
```

[ghstack-poisoned]

desertfire committed:
Update on "[inductor] Allow view to take dtype as its input"


Summary: as title

Test Plan:
```
PYTORCH_TEST_WITH_INDUCTOR=1 python test/test_ops.py -k test_non_standard_bool_values
```

[ghstack-poisoned]

desertfire committed:
Update on "[inductor] Allow view to take dtype as its input"


Summary: as title

Test Plan:
```
PYTORCH_TEST_WITH_INDUCTOR=1 python test/test_ops.py -k test_non_standard_bool_values
```

[ghstack-poisoned]

desertfire committed:
Update on "[inductor] Allow view to take dtype as its input"


Summary: as title

Test Plan:
```
PYTORCH_TEST_WITH_INDUCTOR=1 python test/test_ops.py -k test_non_standard_bool_values
```

[ghstack-poisoned]

ngimel commented on torch/_inductor/utils.py:
```diff
@@ -28,6 +28,21 @@
+    torch.complex32: 4,
+    torch.int64: 8,
+    torch.float64: 8,
+    torch.complex64: 8,
```
should we also add `complex128`?

ngimel commented on test/inductor/test_torchinductor.py:
```diff
@@ -978,6 +978,31 @@ def forward(arg1, arg2):
+    def test_view_dtype(self):
+        def fn(i8_tensor, f16_tensor, f32_tensor, f64_tensor):
+            return (
+                i8_tensor.view(torch.bool),
```
nice test, testing both codegen'd and fallback paths

desertfire commented:
Now I am hitting another issue in `prim` because of lack of support of view with dtype.
```
Traceback (most recent call last):
  File "/scratch/binbao/work/pytorch/torch/_dynamo/utils.py", line 1086, in run_node
    return getattr(args[0], node.target)(*args[1:], **kwargs)
  File "/scratch/binbao/work/pytorch/torch/_subclasses/fake_tensor.py", line 887, in __torch_dispatch__
    r = func(*args, **kwargs)
  File "/scratch/binbao/work/pytorch/torch/_ops.py", line 285, in __call__
    return self._op(*args, **kwargs or {})
  File "/scratch/binbao/work/pytorch/torch/_refs/__init__.py", line 3851, in view
    return _reshape_view_helper(a, *shape, allow_copy=False)
  File "/scratch/binbao/work/pytorch/torch/_refs/__init__.py", line 3052, in _reshape_view_helper
    shape = utils.infer_size(shape, a.numel())
  File "/scratch/binbao/work/pytorch/torch/_prims_common/__init__.py", line 711, in infer_size
    elif d >= 0:
TypeError: '>=' not supported between instances of 'torch.dtype' and 'int'

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/scratch/binbao/work/pytorch/torch/_dynamo/utils.py", line 1050, in get_fake_value
    return wrap_fake_exception(
  File "/scratch/binbao/work/pytorch/torch/_dynamo/utils.py", line 729, in wrap_fake_exception
    return fn()
  File "/scratch/binbao/work/pytorch/torch/_dynamo/utils.py", line 1051, in <lambda>
    lambda: run_node(tx.output, node, args, kwargs, nnmodule)
  File "/scratch/binbao/work/pytorch/torch/_dynamo/utils.py", line 1093, in run_node
    raise RuntimeError(
RuntimeError: Failed running call_method view(*(FakeTensor(FakeTensor(..., device='meta', size=(3, 4), dtype=torch.int8), cuda:0), torch.bool), **{}):
'>=' not supported between instances of 'torch.dtype' and 'int'
(scroll up for backtrace)
```

ngimel commented:
Does this happen before inductor? We probably need meta registration for `view` so that we don't go into refs, cc @SherlockNoMad

desertfire commented:
> Does this happen before inductor? We probably need meta registration for `view` so that we don't go into refs, cc @SherlockNoMad

Yes, it happens to dynamo-eager backend as well.
